models/T2star.py

```diff
@@ -0,0 +1,90 @@
+#!/ur/bin/python
+# -*- coding: utf-8 -*-
+"""
+@author: Kanishka Sharma  
+T2*-mapping signal model-fit  
+2021  
+"""
+
+import numpy as np
+from scipy.optimize import curve_fit
+
+def exp_func(TE,S0,T2star):
+    """ mono-exponential decay function to perform T2*-fitting.  
+
+    Args
+    ----
+    TE (numpy.ndarray): Echo times (TE) in the T2*-mapping sequence as input (independent variable) for the signal model fit.  
+
+    Returns
+    -------
+    S0, T2star(numpy.ndarray): output signal model fit parameters.  
+
+    """
+  
+    return S0*np.exp(-TE/T2star)
+
+
+
+def T2star_fitting(images_to_be_fitted, echo_times):
+    """ curve fit function which returns the fit and fitted params S0 and T2*.
+
+    Args
+    ----
+    images_to_be_fitted (numpy.ndarray):  input image at all time-series (i.e. at each TE time) with shape [x-dim*y-dim, total time-series].    
+    echo_times (list): list containing TE times as input (independent variable) for the signal model fit.      
+
+    Returns
+    -------
+    fit (numpy.ndarray): signal model fit at all time-series with shape [x-dim*y-dim, total time-series].      
+    S0 (numpy.ndarray): fitted parameter 'S0' with shape [x-dim*y-dim].      
+    T2_star (numpy.ndarray): fitted parameter 'T2_star' (ms) with shape [x-dim*y-dim].  
+
+    """
+    lb = [0,10]
+    ub = [np.inf,100]
+    initial_guess = [np.max(images_to_be_fitted),50] 
+    shape = np.shape(images_to_be_fitted)
+    S0 = np.empty(shape[0]) 
+    T2_star = np.empty(shape[0])
+    fit = np.empty(shape)
+
+    for x in range(shape[0]):#pixels (x-dim*y-dim)
+       popt, pcov = curve_fit(exp_func, xdata = echo_times, ydata = images_to_be_fitted[x,:], p0=initial_guess, bounds=(lb,ub), method='trf')
+       S0[x] =  popt[0] 
+       T2_star[x] =  popt[1]
+
+    for x in range(shape[0]): #pixels (x-dim*y-dim)
```

Comment:
This for loop isn't really necessary, one `for x in range(shape[0]):` should be enough for this case, T2 and ADC as well for example. I don't see why these would be in separate and consecutive (no intermediate calculations in between) for loops when the iterator is exactly the same. This is part of the reason why modelling is slow. Putting everything in 1 loop will also make parallelisation (in a future task) easier.

In the case of the DCE there's a case where it makes sense separate loops because there are intermediate calculations between loops that are necessary.